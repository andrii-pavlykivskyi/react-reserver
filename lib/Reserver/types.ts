import { ReactNode } from "react"
import { Dimension } from "../types"
import { CellEventHandler } from "../Cell"

export type ReserverContent = Record<`r${number}c${number}`, ReactNode>

export type GetReserverContent = (
  columnCount: number,
  rowCount: number
) => ReserverContent

type Title = ((count: number) => React.ReactNode) | ReactNode

export type ChildrenProps = {
  rowCount: number
  columnCount: number
  rowTitleWidth: number
  dimension: Dimension
  columnTitleHeight: number
}


// dimension = { height: 20, width: 20 },
// height = 500,
// width = 500,

// columnTitles: getColumnTitles = [],
// rowTitles: getRowTitles = [],
// content: getContent = (() => ({})) as GetReserverContent,
// rowTitleWidth = 0,
// columnTitleHeight = 0,
// dir = 'ltr',


export type ReserverProps = Omit<React.ComponentProps<'div'>, 'content'> & {
  content?: GetReserverContent
  dimension: Dimension
  height: number
  width: number
  rowTitleWidth: number
  rowTitles: Title[]
  columnTitles: Title[]
  columnTitleHeight: number
  columnTitleClassName?: string
  cantonClassName?: string
  rowTitleClassName?: string
  cellClassName?: string

  mouseDownCell: CellEventHandler
  mouseEnterCell: CellEventHandler
  mouseUpCell: CellEventHandler
  mouseDropCell: CellEventHandler
  mouseDragOverCell: CellEventHandler
  pointerDownCell: CellEventHandler
  pointerMoveCell: CellEventHandler
  pointerEnterCell: CellEventHandler
  pointerLeaveCell: CellEventHandler
  pointerUpCell: CellEventHandler
  pointerOverCell: CellEventHandler

  children: (props: ChildrenProps) => ReactNode
}