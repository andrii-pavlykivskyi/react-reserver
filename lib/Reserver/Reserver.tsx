import React, { useEffect } from 'react';
import Head from '../Head';
import Cell from '../Cell';
import { ReserverProps, GetReserverContent } from './types';
import useStyle from '../hooks/useStyle';

const Reserver = React.forwardRef<HTMLDivElement, ReserverProps>(
  (
    {
      dimension = { height: 20, width: 20 },
      columnTitles,
      rowTitles,
      content: getContent = (() => ({})) as GetReserverContent,
      rowTitleWidth = 0,
      columnTitleHeight = 0,
      dir = 'ltr',
      isDragging = false,

      children,
      HeadProps,
      BodyCellProps,
      ...props
    },
    ref
  ) => {
    const setStyle = useStyle(`
    .reserver-cell:hover > div {
      background: #e4e8e9 !important;
    }  
    .peg-drop-place {
      position: relative;
    }
    .peg-drop-place::after {
      content: '';
      inset: 0px;
      background-color: #0e6ba830;
      z-index: 20000;
      position: absolute;
    }
    `);

    useEffect(() => {
      setStyle(`
      .peg-drop-place-first::after {
        border-bottom-left-radius: ${dimension.height}px;
        border-top-left-radius: ${dimension.height}px;
      }
      .peg-drop-place-last::after {
        border-bottom-right-radius: ${dimension.height}px;
        border-top-right-radius: ${dimension.height}px;
      }
      `);
    }, [dimension, setStyle]);

    const columnCount = (columnTitles || []).length;
    const rowCount = (rowTitles || []).length;

    const content = getContent(columnCount, rowCount);

    const minWidth = rowTitleWidth + columnCount * dimension.width;

    return (
      <div
        ref={ref}
        id={props.id}
        className={props.className}
        role="grid"
        {...props}
        style={{ ...props.style, position: 'relative', minWidth }}
      >
        <Head
          columnTitles={columnTitles}
          isVisible={columnTitles.length > 0}
          {...HeadProps}
          columnCount={columnCount}
          height={columnTitleHeight}
          rowTitleWidth={rowTitleWidth}
          dimension={dimension}
          columnTitleClassName={props.columnTitleClassName}
          cantonClassName={props.cantonClassName}
          dir={dir}
        />
        {[...Array(rowCount)].map((_, r) => {
          return (
            <div role="rowgroup" key={r} style={{ height: dimension.height, display: 'flex' }}>
              {dir === 'ltr' && (
                <Cell
                  column={-1}
                  row={r}
                  style={{ display: rowTitles.length > 0 ? 'initial' : 'none' }}
                  dimension={{
                    height: dimension.height,
                    width: rowTitleWidth
                  }}
                  isHeading
                  className={props.rowTitleClassName}
                >
                  {rowTitles[r]}
                </Cell>
              )}
              {[...Array(columnCount)].map((_, c) => {
                return (
                  <Cell
                    {...BodyCellProps}
                    key={`r${r}c${c}`}
                    dimension={dimension}
                    className={props.cellClassName}
                    column={c}
                    row={r}
                    isDragging={isDragging}
                    isHeading={false}
                  >
                    {content[`r${r}c${c}`]}
                  </Cell>
                );
              })}
              {dir === 'rtl' && (
                <Cell
                  column={-1}
                  row={r}
                  isHeading
                  style={{ display: rowTitles.length > 0 ? 'initial' : 'none' }}
                  dimension={{
                    height: dimension.height,
                    width: rowTitleWidth
                  }}
                  className={props.rowTitleClassName}
                >
                  {rowTitles[r]}
                </Cell>
              )}
            </div>
          );
        })}

        <div role="list" style={{ height: '100%', overflow: 'hidden' }}>
          {typeof children === 'function' &&
            children({
              rowCount: rowCount,
              columnCount: columnCount,
              rowTitleWidth: rowTitleWidth,
              dimension: dimension,
              columnTitleHeight:
                columnTitles.length > 0
                  ? columnTitleHeight > 0
                    ? columnTitleHeight
                    : dimension.height
                  : 0
            })}
          {Array.isArray(children) && children}
        </div>
      </div>
    );
  }
);

export default Reserver;
