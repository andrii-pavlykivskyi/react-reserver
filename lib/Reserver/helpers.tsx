import { ReservationItem } from './types';

import moment, { Moment } from 'moment';

type Unit = moment.unitOfTime.DurationConstructor;

export function dateRange(start: string, length: number, unit: Unit): Date[] {
  return [...Array(length)].map((_, i) => {
    return resolveDate(start, i, unit);
  });
}

export function resolveDate(start: string, count: number, unit: Unit) {
  return moment(start).add(count, unit).toDate();
}

export function resolveDateDiff(
  startDate: string | Moment,
  date: string | Moment,
  format = 'DD-MM-YYYY'
) {
  const a = typeof startDate === 'string' ? moment(startDate, format).startOf('day') : startDate;
  const b = typeof date === 'string' ? moment(date, format).startOf('day') : date;

  return b.diff(a, 'days');
}

export function generateRowTitles(
  reservationItems: ReservationItem[],
  dropRowIndex: number | null
) {
  return reservationItems.map((room, index) => {
    return (
      <div style={{ height: '100%' }}>
        <div
          style={{
            height: '100%',
            padding: '3px',
            display: 'flex',
            alignContent: 'center',
            background: dropRowIndex === index ? '#1ca3f9' : room.background
          }}
        >
          <div style={{ fontSize: '13px', marginRight: '10px' }}>{room.number}</div>
          <div>{room.name}</div>
        </div>
      </div>
    );
  });
}
