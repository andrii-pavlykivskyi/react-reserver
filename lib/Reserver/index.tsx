import Reserver from './Reserver';

export default Reserver;
