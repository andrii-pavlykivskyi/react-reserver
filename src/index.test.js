import React from 'react'
import renderer from 'react-test-renderer'
import {
  render,
  screen,
  fireEvent,
  getAllByRole,
  act,
  prettyDOM
} from '@testing-library/react'
import Reserver, {
  Bar,
  useReserver,
  reserverReducer,
  createBar,
  getPosition,
  resizeBar
} from './index'
describe('Reserver', () => {
  test('"Reserver runs', () => {
    render(<Reserver />)
  })
})

function Basic(props) {
  const {
    bars,
    isEditing,
    setIsEditing,
    addBar,
    setBars,
    editBar,
    deleteBar
  } = useReserver(reserverReducer, [])

  return (
    <Reserver
      width={150}
      height={50}
      id='reserver'
      mouseDownCell={(props) => {
        const newbar = createBar(props.dimension, props.cell)
        addBar(newbar)
      }}
      mouseEnterCell={(props) => {
        if (isEditing) {
          const nBars = resizeBar(bars, props)
          setBars(nBars)
        }
      }}
      mouseUpCell={() => {
        const dBars = bars.map((bar) => {
          if (bar.editing) {
            return {
              ...bar,
              editing: false,
              style: { ...bar.style, pointerEvents: 'auto' }
            }
          }
          return bar
        })

        setBars(dBars)
        setIsEditing(false)
      }}
    >
      {bars.map((bar) => {
        return (
          <Bar
            onMouseLeave={(e, props) => {
              editBar({
                ...props,
                style: { ...props.style, background: 'red' }
              })
            }}
            onContextMenu={(e, props) => {
              deleteBar(props)
            }}
            key={bar.id}
            {...bar}
            style={{
              ...bar.style,
              ...getPosition(bar.row, bar.column, bar.dimension)
            }}
          >
            Test
          </Bar>
        )
      })}
    </Reserver>
  )
}

/*
Need to add this to tests
test('headrow receives function ', () => {
  render(
    <Head
      dimension={{width:20,height:20}}
      rowTitleWidth='20px'
      columnTitles={(c) => {
        return [...Array(c)].map((x, i) => {
          return <div key={i}>{i}</div>
        })
      }}
      columnCount={4}
    />
  )
  screen.getByText('1')
})
*/

test('Reserver adds, removes, edits and setsBars', () => {
  const dom = render(<Basic />)

  const square = getAllByRole(dom.container, 'gridcell')

  act(() => {
    fireEvent.mouseDown(square[0])
  })

  const bar = screen.getByText('Test')
  expect(bar).toBeTruthy()

  act(() => {
    fireEvent.mouseOver(square[4])
  })

  const barComp = renderer.create(prettyDOM(bar)).toJSON()
  expect(barComp).toMatchSnapshot()

  const newBar = screen.getByText('Test')
  act(() => {
    fireEvent.mouseUp(square[4])
  })
  const barwithauto = renderer.create(prettyDOM(newBar)).toJSON()
  expect(barwithauto).toMatchSnapshot()

  act(() => {
    fireEvent.mouseEnter(bar)
    fireEvent.mouseLeave(bar)
  })
  const redBar = screen.getByText('Test')

  act(() => {
    fireEvent.contextMenu(bar)
  })

  const redbarsnap = renderer.create(prettyDOM(redBar)).toJSON()
  expect(redbarsnap).toMatchSnapshot()

  const nobar = screen.queryAllByText('Test')
  expect(nobar).toEqual([])

  // screen.debug(nobar);
})
