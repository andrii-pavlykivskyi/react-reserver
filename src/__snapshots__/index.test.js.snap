// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reserver adds, removes, edits and setsBars 1`] = `
"[36m<div[39m
  [33mid[39m=[32m\\"ojWT7nwV0qeqbmA\\"[39m
  [33mrole[39m=[32m\\"listitem\\"[39m
  [33mstyle[39m=[32m\\"left: 0px; top: 0px; pointer-events: none; background: rgb(14, 107, 168); display: flex; position: absolute; z-index: 100;\\"[39m
[36m>[39m
  [36m<div[39m
    [33mstyle[39m=[32m\\"width: 20px; height: 20px; pointer-events: none;\\"[39m
  [36m/>[39m
  [0mTest[0m
[36m</div>[39m"
`;

exports[`Reserver adds, removes, edits and setsBars 2`] = `
"[36m<div[39m
  [33mid[39m=[32m\\"ojWT7nwV0qeqbmA\\"[39m
  [33mrole[39m=[32m\\"listitem\\"[39m
  [33mstyle[39m=[32m\\"left: 0px; top: 0px; pointer-events: auto; background: rgb(14, 107, 168); display: flex; position: absolute; z-index: 100;\\"[39m
[36m>[39m
  [36m<div[39m
    [33mstyle[39m=[32m\\"width: 20px; height: 20px; pointer-events: auto;\\"[39m
  [36m/>[39m
  [0mTest[0m
[36m</div>[39m"
`;

exports[`Reserver adds, removes, edits and setsBars 3`] = `
"[36m<div[39m
  [33mid[39m=[32m\\"ojWT7nwV0qeqbmA\\"[39m
  [33mrole[39m=[32m\\"listitem\\"[39m
  [33mstyle[39m=[32m\\"left: 0px; top: 0px; pointer-events: auto; display: flex; position: absolute; z-index: 100; background: red;\\"[39m
[36m>[39m
  [36m<div[39m
    [33mstyle[39m=[32m\\"width: 20px; height: 20px; pointer-events: auto;\\"[39m
  [36m/>[39m
  [0mTest[0m
[36m</div>[39m"
`;
